### ISSUE TITLE
"process" dependency in UMD bundle breaks usage in browsers.

### ISSUE THREAD

**Issue by mbostock** [nteract/semiotic#498]
The [current UMD bundle](https://unpkg.com/semiotic@1.20.2/dist/semiotic.min.js) has a dependency on the `process` module:

```js
require("process")
```

Probably the code expects this to be Node’s [process built-in](https://nodejs.org/api/process.html), but on the web, this resolves to a browser shim:

https://www.npmjs.com/package/process

And this browser shim doesn’t support UMD, which breaks. We can workaround this issue by creating an empty object for `process` as follows using [require.alias](https://github.com/d3/d3-require/blob/master/README.md#require_alias):

```js
r = require.alias({
  "process": {},
  "react": "react@16/umd/react.production.min.js",
  "react-dom": "react-dom@16/umd/react-dom.production.min.js",
  "semiotic": "semiotic@1"
})
```

But my guess is that `process` should be declared a global in your [rollup.config](https://github.com/nteract/semiotic/blob/master/rollup.config.js) to avoid the unnecessary require of an unintended module.

**Comment by emeeks** [comment]
I tried to move this up to `globals` but then it doesn't build so I'll see if there's another way to declare it as a global and see how to make it work but if you have any pointers that would be much appreciated--rollup is not exactly the best when it comes to figuring these things out.

**Comment by mbostock** [comment]
Did you move it from externals to globals, or did you just add it to globals? I would guess the latter.

**Comment by emeeks** [comment]
You're right, I originally moved it from externals to globals. I didn't realize you could have it in both places. Seems to build fine like that. Is there a straightforward way to see if the UMD bundle it produces is working without publishing it?

**Comment by mbostock** [comment]
You could use the “require local file” technique shown here:

https://observablehq.com/@mbostock/reading-local-files

**Comment by mbostock** [comment]
I tested it locally, and yeah, we’d need to move the process from externals to globals, not have it in both places. But when doing that, the code then tries to call `process.nextTick` somewhere, so more work is required here and I don’t know enough to suggest a fix.

**Comment by emeeks** [comment]
Yeah, that was the problem I ran into. I'll see if I can figure out a way to get it not to break on that.

**Comment by emeeks** [comment]
So I think I have a working build with it as a global. The problem is that the only two ways that I've tested whether or not it's working is via a puppeteer based test and also Codepen, but both of those were currently with the other setup so is there some way I can test to see if it's working in the use case(s) that you're using?

**Comment by mbostock** [comment]
The require implementation we use in Observable for UMD is d3-require:

https://github.com/d3/d3-require

If you’ve got browser-based tests already, you should be able to do something like this to test:

https://github.com/d3/d3-require/blob/master/test/test-dependency.html

Just use a relative path `./semiotic.js` to load your file (and then optionally do the require.alias stuff for React).

**Comment by emeeks** [comment]
Thanks for the guidance, it looks like I got it working with d3-require with the rollup settings. I'm going to publish a patch as soon as I've got some other bits cleaned up so if you get a chance to take a look, great, but otherwise I'm feeling good.

**Comment by emeeks** [comment]
This should be fixed in 1.20.3. Please feel free to reopen this and let me know if there's still an issue.

**Comment by mbostock** [comment]
Works great! Thanks for the fix.
